proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,1066 @@
+for a given key to match any resource variant that does not specify that
+key at all.  This allows constraints for new keys to be added on clients
+before the corresponding keys are added on the resources on the server, but
+it does introduce some additional ambiguity into the matching.  For example,
+let's say that the server has the following two variants of a resource:
+- `{env=prod}`
+- `{env=prod, version=v1}`
+
+Now consider what happens if a client subscribes with the following
+constraints:
+
+```textproto
+{key_constraints:[
+  {key:"env" value:{
+    constraints:[{value:"prod"}]
+  }},
+  {key:"version" value:{
+    constraints:[{value:"v1"}]
+  }}
+]}
+```
+
+These constraints can match either of the above variants of the resource.
+This situation can be avoided by establishing a best practice that all
+variants of a given resource must have the same set of keys.
+
+There is still a possible ambiguity that can occur if a server adds
+multiple variants of a new key that clients are not yet sending.
+For example, let's say that the server has the following two variants
+of a resource:
+- `{env=prod, version=v1}`
+- `{env=prod, version=v2}`
+
+Consider what happens if a client subscribes with the following constraints:
+
+```textproto
+{key_constraints:[
+  {key:"env" value:{
+    constraints:[{value:"prod"}]
+  }}
+]}
+```
+
+These constraints can match either variant of the above resource.
+This can be avoided by establishing a best practice of not adding multiple
+variants of a new parameter until clients are sending the new parameter.
+However, if this does happen, the cache implementation is free to pick
+one of the variants at random.
+
+So, the expected order of changes for this kind of transition would be:
+1. Change clients to start sending a constraint for `version=v1`.
+2. Add the dynamic parameter `version=v1` to all existing resources.
+3. Create new variants of each resource with `version=v2`.
+4. Change the desired set of clients to send a constraint for
+   `version=v2` instead of `version=v1`.
+
+##### Alternatives Considered
+
+We could avoid much of the matching ambiguity described above by saying that
+a set of constraints must specify all keys present on the resource in order
+to match.  However, this would mean that if the client starts subscribing
+with a constraint for the new key before the corresponding key is added on
+the resources on the server, then it will fail to match the existing resources.
+In other words, the process would be:
+
+1. Add a variant of all resources on the server side with `version=v1`
+   (in addition to all existing dynamic parameters).
+2. Change clients to start sending constraints with the new key.
+3. When all clients are updated, remove the resource variants that do
+   *not* have the new key.
+
+This will effectively require adding new keys on the server side first,
+which seems like a large burden on users.  It also seems fairly tricky
```

Comment:
Why is this so bad? We basically need to roll out somewhere first.

Reply:
If we think users can deal with this, I'd actually prefer this approach, because it would basically eliminate the ambiguity problem.  But from what @fuqianggao has said, I believe that one of the common use-cases is the one I mentioned above, where users divide their clients into categories based on some key, even if they initially don't configure those sets of clients differently from each other, so that they can easily change the config of a subset of their clients later simply by changing the configuration reesources in the control plane.  I'm not sure how to support that use-case with requiring all keys to exist.